refactor: `alt.themes` -> `alt.theme`

dangotbanned opened the pull request:
Will close #3610, #3607

# Description
This PR aims to consolidate theme related functionality behind a single namespace.
Plans & discussion have been collected in:
- https://github.com/vega/altair/issues/3610#issue-2545031537
- https://github.com/vega/altair/issues/3610#issuecomment-2374650764
- https://github.com/vega/altair/issues/3610#issuecomment-2378029196

# Implementation
I would recommend reading https://github.com/vega/altair/pull/3618#discussion_r1780006759 first, as an unexpected issue arose.
Solving this contributed **+1400** additional lines, of which **85%** is generated code in (https://github.com/vega/altair/pull/3618/files#diff-3426a297d4278b99a32d85b11c425f5479dab750a63ba4ff77ba75870946cadf), (https://github.com/vega/altair/pull/3618/files#diff-7ed2cae75fa3376b19115773b5d58bd76f12c6e10a0ed7a03b22e098842b7b1f)

# Functional Changes
- https://github.com/vega/altair/issues/3610#issuecomment-2385432948

# Tasks
- [x] Add deprecation handling `alt.__init__.__getattr__`
- [x] Add new module `alt.theme.py`
- [x] `@alt.register_theme` -> `@alt.theme.register`
- [x] Add `alt.theme.enable`
- [x] [feat: Support `alt.theme.(active|options)`](https://github.com/vega/altair/pull/3618/commits/58da996dbe19f3750159713fa0b73ef44c8455f1)
- [x] API Reference [thread](https://github.com/vega/altair/pull/3618#discussion_r1780130320)
	- [docs: Adds alt.theme to API Reference](https://github.com/vega/altair/pull/3618/commits/217d7fb181275999ae3f95a6215a6aeb53825353)
	- [Screenshots](https://github.com/vega/altair/pull/3618#discussion_r1780184816)
- [x] Add missing docs
	- [x] [`generate_schema__init__`](https://github.com/vega/altair/pull/3618#discussion_r1780132853) -> [`48e44c0` (#3618)](https://github.com/vega/altair/pull/3618/commits/48e44c0e4adbc740f7fe3c7aa35cf02072e7802c)
- [x] Finalize deprecation message [thread](https://github.com/vega/altair/pull/3618#discussion_r1780150958)
	- [docs: Provide API ref link in warning](https://github.com/vega/altair/pull/3618/commits/c8033b2cdb7939eda9b5a2d839d008a5fa8c59ba)
- [x] User Guide [thread](https://github.com/vega/altair/pull/3618#discussion_r1780128126)
	- [docs: Update User Guide section](https://github.com/vega/altair/pull/3618/commits/e11449322d14775cdf087ce40f590acd9b232f1d)

dangotbanned commented on altair/vegalite/v5/__init__.py:
# FIXME
**This file is currently blocking**

AFAIK `altair.vegalite.v5.theme` cannot be moved or renamed without breaking

https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/vegalite/v5/theme.py#L52-L56

This itself is not an issue, but a side effect of where `theme.py` is defined is that it is also accesible as `alt.theme` already.

Since `altair.vegalite.v5.__init__` does not declare an `__all__`, when the level above uses `from .v5 import *` all of the module names propagate

https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/vegalite/__init__.py#L2

This results in the top-level having modules accessible which were never explicitly imported

https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/__init__.py#L600

https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/__init__.py#L629

---

# Complications
Declaring an `__all__` is complicated by the following:

## `*` imports from both generated + regular files

https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/vegalite/v5/__init__.py#L4
https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/vegalite/v5/__init__.py#L23

## Ambiguities around similar imports
https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/vegalite/api.py#L2
https://github.com/vega/altair/blob/cabf1e6428bab107a26adc6c51993d3b2c5bf6f0/altair/vegalite/schema.py#L4

## `mypy` vs `pyright`
I was able to resolve this for both runtime **and** statically with `pyright` (`pylance`).
Following the docs, I used one of the [accepted idioms](https://microsoft.github.io/pyright/#/typed-libraries?id=library-interface).

I chose this one specifically because I have seen it used in https://github.com/python/cpython/blob/3f27153e077d7e9448e2f081275931968b40cc74/Lib/asyncio/__init__.py

<details><summary><code>altair.vegalite.v5.__init__.py</code></summary>
<p>


```py
# ruff: noqa: F401, F403
from altair.expr.core import datum
from altair.vegalite.v5 import api, compiler, schema
from altair.vegalite.v5.api import *
from altair.vegalite.v5.compiler import vegalite_compilers
from altair.vegalite.v5.data import (
    MaxRowsError,
    data_transformers,
    default_data_transformer,
    limit_rows,
    sample,
    to_csv,
    to_json,
    to_values,
)
from altair.vegalite.v5.display import (
    VEGA_VERSION,
    VEGAEMBED_VERSION,
    VEGALITE_VERSION,
    VegaLite,
    renderers,
)
from altair.vegalite.v5.schema import *

__all__ = [
    "VEGAEMBED_VERSION",
    "VEGALITE_VERSION",
    "VEGA_VERSION",
    "MaxRowsError",
    "VegaLite",
    "api",
    "compiler",
    "data_transformers",
    "datum",
    "default_data_transformer",
    "limit_rows",
    "renderers",
    "sample",
    "schema",
    "to_csv",
    "to_json",
    "to_values",
    "vegalite_compilers",
]

__all__ += api.__all__
__all__ += schema.__all__

```

</p>
</details> 

However, despite all tests working as expected - `mypy` no longer recognises any imports.

# Solution(s)
...

dangotbanned replied:
I've now got this locally:

<details><summary>build(typing): Generate <code>altair.vegalite.v5.schema.__init__.__all__</code></summary>
<p>

![image](https://github.com/user-attachments/assets/d37eea51-08b4-4aa1-90c7-6ea12dc9bbf8)


</p>
</details> 

Going to try and work out a solution for `altair.vegalite.v5.__init__.__all__` next.

Hopefully will be able to start pushing the rest of the changes afterwards

dangotbanned replied:
Fixed in:
- [build(typing): Generate altair.vegalite.v5.schema.__init__.__all__](https://github.com/vega/altair/pull/3618/commits/b1be6335afd56c8ea1f6dd0905775a9255268db8)
- [build: Generate altair.vegalite.v5.__init__.__all__](https://github.com/vega/altair/pull/3618/commits/0bd8f755651a2d00c4695907b6f7eceb87bc0bf2)

dangotbanned replied:
Another benefit of this fix was that `tools/generate_api_docs.py` no longer requires [these ignore comments](https://github.com/vega/altair/commit/f7fe38bd70fadbb4309320567263fd9762bf71e0)

dangotbanned commented on doc/user_guide/customization.rst:
```diff
@@ -710,6 +710,9 @@ outside the chart itself; For example, the container may be a ``<div>`` element
 
 Chart Themes
 ------------
+..
+    _comment: First mention of alt.themes
+
```
# TODO
- Replace all `alt.themes` references
- Plan new content
- Add a callout stating when and why the change was made
- Possibly link to related issues

dangotbanned commented on tools/generate_api_docs.py:
```diff
@@ -134,8 +134,13 @@ def type_hints() -> list[str]:
     return sorted(s for s in iter_objects(alt.typing) if s in alt.typing.__all__)
 
 
+# TODO: Currently only the `TypedDict`(s) are visible (only via `alt.typing.___`)
+# Related: https://github.com/vega/altair/issues/3607
+def theme() -> list[str]: ...  # type: ignore[empty-body]
+
```
# TODO
- Add section for API reference

dangotbanned commented on tools/generate_schema_wrapper.py:
```diff
@@ -1001,6 +1018,32 @@ def generate_vegalite_config_mixin(fp: Path, /) -> str:
     return "\n".join(code)
 
 
+def generate_schema__init__(
+    version: str,
+    *modules: str,
+    package_name: str = "altair.vegalite.{0}.schema",
+    expand: dict[Path, ModuleDef[Any]] | None = None,
+) -> Iterator[str]:
+    # NOTE: `expand`
+    # - Should run after generating `core`, `channels`
+    # - Only needed for `mypy`, the default works at runtime
+    package_name = package_name.format(version.split(".")[0])
```
# TODO
- Add docstring

dangotbanned commented on altair/vegalite/v5/theme.py:
```diff
@@ -28,11 +20,47 @@
 VEGA_THEMES: list[LiteralString] = list(get_args(VegaThemes))
 
 
+# HACK: See for `LiteralString` requirement in `name`
+# https://github.com/vega/altair/pull/3526#discussion_r1743350127
+class ThemeRegistry(PluginRegistry[Plugin[ThemeConfig], ThemeConfig]):
+    def enable(
+        self,
+        name: LiteralString | AltairThemes | VegaThemes | None = None,
+        **options: Any,
+    ) -> PluginEnabler:
+        """
```
# NOTE
Previously the only thing in https://github.com/vega/altair/pull/3618/files#diff-43d26f1359b1c241f9f632a56cbc8cef42f2fee40941e284620ebefc3de0b25e

dangotbanned commented on altair/theme.py:
```diff
@@ -0,0 +1,266 @@
+from functools import wraps as _wraps
+from typing import TYPE_CHECKING
+
+from altair.vegalite.v5.schema._config import (
+    AreaConfigKwds,
+    AutoSizeParamsKwds,
+    AxisConfigKwds,
+    AxisResolveMapKwds,
+    BarConfigKwds,
+    BindCheckboxKwds,
+    BindDirectKwds,
+    BindInputKwds,
+    BindRadioSelectKwds,
```
# NOTE
- I've manually copied these over for now.
- Would prefer this to using a `*` import
- May explore managing the declaration of this in a similar way to the `__all__` definition

dangotbanned commented on altair/__init__.py:
```diff
@@ -653,10 +653,29 @@ def __dir__():
+def __getattr__(name: str) -> _Any:
+    from altair.utils.deprecation import deprecated_warn
+
+    if name == "themes":
+        deprecated_warn(
+            "Most of the `ThemeRegistry` API is accessible via `altair.theme`.\n"
+            "See the updated User Guide for further details:\n"
+            "https://altair-viz.github.io/user_guide/customization.html#chart-themes",
```
Placeholder message, revisit before merging

dangotbanned commented on altair/__init__.py:
```diff
@@ -653,10 +653,29 @@ def __dir__():
+def __getattr__(name: str) -> _Any:
+    from altair.utils.deprecation import deprecated_warn
+
+    if name == "themes":
```
Solution is fine currently, but as soon as we start deprecating other imports it would make sense to generalize

dangotbanned commented on altair/__init__.py:
```diff
@@ -653,10 +653,29 @@ def __dir__():
+            "https://altair-viz.github.io/user_guide/customization.html#chart-themes",
+            version="5.5.0",
+            alternative="altair.theme.themes",
+            stacklevel=3,
```
While I was testing, the default I'd set of 2 didn't report any warnings.

Not sure if that is related to `__getattr__` itself, but noting it anyway

dangotbanned commented on altair/vegalite/v5/schema/_config.py:
```diff
@@ -93,7 +93,7 @@
 class AreaConfigKwds(TypedDict, total=False):
     """
-    :class:`AreaConfig` ``TypedDict`` wrapper.
+    :class:`altair.AreaConfig` ``TypedDict`` wrapper.
```
Previously these didn't work as links in the API reference.

Now they do

![image](https://github.com/user-attachments/assets/46985e0b-0a2a-418c-a323-52a702131b41)

dangotbanned commented on altair/theme.py:
```diff
@@ -0,0 +1,266 @@
+    return themes.register(name, None)
+
+
+enable = themes.enable
+get = themes.get
+names = themes.names
```
# Note
Originally planned for this to be a temporary solution.

Quite surpised at how well aliasing worked:

<details>
<summary>Docstring</summary>

![image](https://github.com/user-attachments/assets/64ee48bb-8dca-4236-9506-a431ede9227f)

</details>

<details><summary>API Reference</summary>
<p>

![image](https://github.com/user-attachments/assets/2f818eb8-700c-4f76-b0d6-f4e9589ed6c0)

</p>
</details> 

<details><summary>Indistinguishable from an actual function</summary>
<p>

![image](https://github.com/user-attachments/assets/d833d0aa-f513-461b-a4fc-9f09cd9e64ec)

</p>
</details>

dangotbanned commented on doc/user_guide/api.rst:
```diff
@@ -164,6 +170,95 @@ API Functions
    vconcat
    when
 
+.. _api-theme:
+
+Theme
+-----
+.. currentmodule:: altair.theme
+
+.. autosummary::
+   :toctree: generated/theme/
+   :nosignatures:
+
+   enable
+   get
+   names
+   register
+   themes
+   unregister
+   ThemeConfig
+   AreaConfigKwds
+   AutoSizeParamsKwds
```
# Preview
I think this order makes the most sense.

- Functions will be the first thing you need
- The top level config is clearly visible
- Then further config via `...Kwds`.

![image](https://github.com/user-attachments/assets/49b6963d-96b5-419a-b7a5-6296ea44a82d)

dangotbanned commented on doc/user_guide/customization.rst:
```diff
@@ -710,18 +710,20 @@ outside the chart itself; For example, the container may be a ``<div>`` element
 ------------
-..
-    _comment: First mention of alt.themes
+.. note::
+
+   This material was changed considerably with the release of Altair ``5.5.0``.
```
# Note
Trying to keep this PR focused to make for an easier review. 

I am planning to develop this section further (#3519), but would like to do that in follow-up PRs.

(All still prior to `v5.5.0`)

dangotbanned replied:
Opened #3645 to track this, so marking as resolved

dangotbanned commented on altair/__init__.py:
```diff
@@ -653,10 +651,30 @@ def __dir__():
+    from altair.utils.deprecation import deprecated_warn
+
+    if name == "themes":
+        deprecated_warn(
+            "Most of the `ThemeRegistry` API is accessible via `altair.theme`.\n"
+            "See the updated User Guide for further details:\n"
+            "    https://altair-viz.github.io/user_guide/api.html#theme\n"
+            "    https://altair-viz.github.io/user_guide/customization.html#chart-themes",
```
`alt.themes` deprecation warning message

dangotbanned commented on altair/theme.py:
```diff
@@ -0,0 +1,290 @@
+    return decorate
+
+
+def unregister(name: LiteralString) -> Plugin[ThemeConfig] | None:
+    """
+    Remove and return a previously registered theme.
+
+    Parameters
+    ----------
+    name
+        Unique name assigned in ``alt.theme.themes``.
+    """
+    return themes.register(name, None)
```
# Note
- More explicit than being part of `themes.register`
- This behavior is **not part of `@alt.theme.register`**
- It wouldn't make sense decorating a function with it

mattijn commented:
First, thanks for all the work and the short and long explanations in your comments!

I found it a bit hard to explain my thoughts in text alone, so let me add some contextual diagrams to it. I aim to distinguish three options:

1. _User experience today_ 
2. _User experience of this PR_
3. _Maintaining familiarity while modernizing code_
 
I feel option 3 hasn't been explored enough, though it seems like a valid approach. I apologize, if I've misunderstood your previous comments if you did already discussed this. In that case, could you please direct me to where this option was discussed.

#### 1. User experience today

Given the following code snippet
```python
import altair as alt
from vega_datasets import data

source = data.cars.url
alt.themes.enable('dark')

alt.Chart(source).mark_tick().encode(
    x="Horsepower:Q",
)
```
> ![image](https://github.com/user-attachments/assets/5140049c-4dbe-4861-b563-8708d06de3f5)

This generates a dark-themed chart without a `DeprecationWarning`. The chart is created using existing legacy code, as shown in the diagram below:

```mermaid
graph TD
    subgraph A["alt.themes.enable('dark')"]
        D["Chart"]
    end
    C["Legacy Code"]    
    
    A <--> C
    style A fill:#f9f,stroke:#333,stroke-width:2px
    style C fill:#b3e0ff,stroke:#333,stroke-width:2px
    style D fill:#ffcc00,stroke:#333,stroke-width:2px
    
    linkStyle 0 stroke:#FF5733, stroke-width:2px; 
```

#### 2. User experience of this PR
From my understanding, this PR will lead to the following behavioral changes for users:
```python
import altair as alt
from vega_datasets import data

source = data.cars.url
alt.themes.enable('dark')

alt.Chart(source).mark_tick().encode(
    x="Horsepower:Q",
)
```
> ```bash
> C:\Users\Hoek\AppData\Local\Temp\ipykernel_28532\3399117876.py:5: AltairDeprecationWarning: Deprecated in `altair=5.5.0`. Use altair.theme.themes instead.
> Most of the `ThemeRegistry` API is accessible via `altair.theme`.
> See the updated User Guide for further details:
>     https://altair-viz.github.io/user_guide/api.html#theme
>     https://altair-viz.github.io/user_guide/customization.html#chart-themes
>   alt.themes.enable('dark')
> ```
> 
> ![image](https://github.com/user-attachments/assets/5140049c-4dbe-4861-b563-8708d06de3f5)

The underlying changes are as follows:

```mermaid
graph TD
    subgraph A["alt.themes.enable('dark')"]
        D["Deprecation Warning + Chart"]
    end
    subgraph E["alt.theme.enable('dark')"]
        F["Chart"]
    end
    B["Modern Code"]
    C["Legacy Code"]
    
    A <--> B
    E <--> B
    B <--> C
    
    style A fill:#f9f,stroke:#333,stroke-width:2px
    style B fill:#f2f2f2,stroke:#333,stroke-width:2px, stroke-dasharray: 5 5
    style C fill:#b3e0ff,stroke:#333,stroke-width:2px
    style D fill:#ffcc00,stroke:#333,stroke-width:2px
    style F fill:#ffcc00,stroke:#333,stroke-width:2px    
    style E fill:#f9f,stroke:#333,stroke-width:2px
    
    linkStyle 0 stroke:#FF5733, stroke-width:2px; 
    linkStyle 1 stroke:#33FF57, stroke-width:2px; 
    linkStyle 2 stroke:#FF5733, stroke-width:2px; 
```
The arrow colors indicate the following: using `alt.themes` (red arrows) will trigger a `DeprecationWarning` and interact with legacy code, while using `alt.theme` (green arrow) does not affect legacy code and resolves everything with the proposed modern code.

#### 3. Maintaining familiarity while modernizing code

I’m curious whether this could be a feasible option as well. It’s more rigorous than consolidating:

```python
import altair as alt
from vega_datasets import data

source = data.cars.url
alt.themes.enable('dark')

alt.Chart(source).mark_tick().encode(
    x="Horsepower:Q",
)
```
> ![image](https://github.com/user-attachments/assets/5140049c-4dbe-4861-b563-8708d06de3f5)

The user experience remains the same as today, with only `alt.themes` available, but the following happens under the hood:
```mermaid
graph TD
    subgraph A["alt.themes.enable('dark')"]
        D["Chart"]
    end
    B["Modern Code"]    
    
    A <--> B
    style A fill:#f9f,stroke:#333,stroke-width:2px
    style B fill:#f2f2f2,stroke:#333,stroke-width:2px, stroke-dasharray: 5 5    
    style D fill:#ffcc00,stroke:#333,stroke-width:2px
    
    linkStyle 0 stroke:#33FF57, stroke-width:2px; 
```

In this case, legacy code is replaced with modern code while maintaining the existing user experience.

Based on the Sourcegraph statistics (great that you brought this up in https://github.com/vega/altair/issues/3610#issuecomment-2385432948), I see the following current usages:

```python
alt.themes.enable()
alt.themes.register()
alt.themes.names()
alt.themes.active
```

If these functions continue to work with a modernized codebase, I prefer this option. By "working," I mean that the final output of the visualizations remains the same as before. I completely understand that the implementations of these functions may differ, and that is perfectly fine.

I hope I’ve accurately interpreted your comments to make me think that this option hasn’t been discussed yet. Anyway, thank you again for your efforts on this!

dangotbanned commented:
> First, thanks for all the work and the short and long explanations in your comments!

Thanks @mattijn

> I feel option 3 hasn't been explored enough, though it seems like a valid approach. I apologize, if I've misunderstood your previous comments if you did already discussed this. In that case, could you please direct me to where this option was discussed.

I hope this can serve as an overview for you

# Prior Comments
> I hope I’ve accurately interpreted your comments to make me think that this option hasn’t been discussed yet.

The central issue is the difference between:
- `alt.theme.register`
- `alt.themes.register`

I discussed this in:
- Section on [`alt.themes.register`](https://github.com/vega/altair/issues/3610#issuecomment-2374650764)
- Bullets in [Functional Changes - `alt.themes.register`](https://github.com/vega/altair/issues/3610#issuecomment-2385432948)

I also think [PEP 318 – Decorators for Functions and Methods](https://peps.python.org/pep-0318/) might be helpful in understanding what is going on here

<details><summary><code>@alt.theme.register</code></summary>
<p>

*Moved in this PR to [alt.theme](https://github.com/vega/altair/blob/d3dccab80518a32a3ba9471c101f1d080e907f30/altair/theme.py#L182)*, but the same as

https://github.com/vega/altair/blob/02ad17d0f1f71ea5125a7c6c9b43a61fd80c4567/altair/vegalite/v5/theme.py#L137-L148

</p>
</details> 

# Real-world Usage

Looking at the [Sourcegraph](https://sourcegraph.com/search?q=context:global+alt.themes.+not+repo:altair+lang:Python+count:11000&patternType=keyword&sm=0) again, every usage of `alt.themes.register` follows this pattern:


```py
import altair as alt

# Define a theme function
# - Variant 1
def custom_theme():
    return {"height": 400, "width": 700}

# - Variant 2 (See Note)
custom_theme = lambda: {"height": 400, "width": 700}

# Register and then **immediately** enable theme
alt.themes.register("theme_name", custom_theme)
alt.themes.enable("theme_name")
```

These can just be replaced, and the result is easier to maintain - since everything is defined in one place:
```py
import altair as alt

@alt.theme.register("theme_name", enable=True)
def custom_theme():
    return {"height": 400, "width": 700}
```

<details>
<summary>Variant 2 Note</summary>
<p>

Technically the sole `lambda` case is different

https://github.com/vega/dash-vega-components/blob/c3e8cae873580bc7a52bc01daea1f27a7df02b8b/example_app.py#L13-L17

Not too concerned with maintaining compatibility for an example in a related `vega` project, as we can just update that.

If this *had* to be a `lambda` you could do:

```py
from tests import slow

@slow
def test_theme_remote() -> None:
    import altair as alt  # noqa: I001
    from urllib.request import urlopen
    import json

    URL = "https://gist.githubusercontent.com/binste/b4042fa76a89d72d45cbbb9355ec6906/raw/e36f79d722bcd9dd954389b1753a2d4a18113227/altair_theme.json"
    with urlopen(URL) as response:
        custom_theme = json.load(response)

    alt.theme.register("remote_binste", enable=True)(lambda: custom_theme)
    assert alt.theme.active == "remote_binste"
```

</p>
</details>

## Non-usage

I found **no usage** of:
```py
alt.themes.register("theme_name", None)
```
This case was the *main focus* of my discussion in https://github.com/vega/altair/issues/3610#issuecomment-2374650764.

### Note

Unsure if you've seen, but I opened https://github.com/vega/altair/issues/3619.

My takeaway was that the current behavior may produce unexpected results.

# Deprecation

I still stand by my prior [conclusion](https://github.com/vega/altair/issues/3610#issuecomment-2374650764)

> I really do think it would be simpler to start fresh with a new namespace, without any additional baggage.
> ...
> We could also be sure that this wouldn't change the meaning of existing user code.
> A deprecation warning can always be ignored for those determined to still use `alt.themes`

I'd add that situations like this are exactly what the [Versioning Policy](https://altair-viz.github.io/about/versioning.html) is there for.

However, by the definition given there `alt.themes` is **not** part of the [Public API](https://altair-viz.github.io/about/versioning.html#public-api).
This PR **adds** `alt.theme` to the Public API, and treats `alt.themes` **as-if** it were previously - by providing a deprecation path.

mattijn commented:
Thanks for comment. Than we have discussed all my remarks. Thank you for your patience!

Can we enhance the `DeprecationWarning` to smoothen the transition?
Something as such:
```python
deprecated_warn(
    "Deprecated in `altair=5.5.0`. Please update your code to use the new `altair.theme` API.\n\n"
    "Update path:\n"
    "Old method (deprecated):\n"
    "    import altair as alt\n"
    "    def custom_theme():\n"
    "        return {'height': 400, 'width': 700}\n\n"
    "    alt.themes.register('theme_name', custom_theme)\n\n"
    "New method (updated):\n"
    "    @alt.theme.register('theme_name', enable=True)\n"
    "    def custom_theme():\n"
    "        return {'height': 400, 'width': 700}\n\n"
    "    alt.theme.enable('theme_name')\n\n"
    "The `altair.theme.themes` method is provided for backward compatibility but should be avoided in new code.\n"
    "For best practices, adopt the new `altair.theme` API.\n"
    "See the updated User Guide for further details:\n"
    "    https://altair-viz.github.io/user_guide/api.html#theme\n"
    "    https://altair-viz.github.io/user_guide/customization.html#chart-themes\n",
    version="5.5.0",
    alternative="alt.theme",
    stacklevel=3,
)
```
I was confused that the suggested alternative was `alt.theme.themes` instead of `alt.theme`.

dangotbanned commented:
> Thanks for comment. Than we have discussed all my remarks. Thank you for your patience!

No problem, thanks again @mattijn 

> Can we enhance the `DeprecationWarning` to smoothen the transition? Something as such:

<details><summary>Quoted code block</summary>
<p>

```py
deprecated_warn(
    "Deprecated in `altair=5.5.0`. Please update your code to use the new `altair.theme` API.\n\n"
    "Update path:\n"
    "Old method (deprecated):\n"
    "    import altair as alt\n"
    "    def custom_theme():\n"
    "        return {'height': 400, 'width': 700}\n\n"
    "    alt.themes.register('theme_name', custom_theme)\n\n"
    "New method (updated):\n"
    "    @alt.theme.register('theme_name', enable=True)\n"
    "    def custom_theme():\n"
    "        return {'height': 400, 'width': 700}\n\n"
    "    alt.theme.enable('theme_name')\n\n"
    "The `altair.theme.themes` method is provided for backward compatibility but should be avoided in new code.\n"
    "For best practices, adopt the new `altair.theme` API.\n"
    "See the updated User Guide for further details:\n"
    "    https://altair-viz.github.io/user_guide/api.html#theme\n"
    "    https://altair-viz.github.io/user_guide/customization.html#chart-themes\n",
    version="5.5.0",
    alternative="alt.theme",
    stacklevel=3,
)
```

</p>
</details> 


> I was confused that the suggested alternative was `alt.theme.themes` instead of `alt.theme`.

Ah yeah, I had a related thought in https://github.com/vega/altair/issues/3610#issuecomment-2385487732 but I'd forgotten the observation.

For reference, this is the format string used
https://github.com/vega/altair/blob/02ad17d0f1f71ea5125a7c6c9b43a61fd80c4567/altair/utils/deprecation.py#L23-L32

For the current warning that corresponds to 
```py
C:/Users/../ipykernel_17380/2846531205.py:1: AltairDeprecationWarning: """Deprecated in `altair={version='5.5.0'}`. Use {alternative='altair.theme.themes'} instead.
{message='Most of the `ThemeRegistry` API is accessible via `altair.theme`.
See the updated User Guide for further details:
    https://altair-viz.github.io/user_guide/api.html#theme
    https://altair-viz.github.io/user_guide/customization.html#chart-themes'}"""
  alt.themes # stacklevel=3 caller
```
---

I'll follow this comment up with some options, just wanted to set the baseline to build on

dangotbanned commented:
Continuation of https://github.com/vega/altair/pull/3618#issuecomment-2408506977

Some of these ideas overlap or may remove the need for others.

# Ideas

1. Change `alternative="altair.theme.themes"` -> `alternative="altair.theme"` in [deprecated_warn](https://github.com/vega/altair/pull/3618#discussion_r1780153381)
	- I think this makes sense in all cases. 
	- My original decision was made prior to supporting `alt.theme.(active|options)` [#3610-comment-6](https://github.com/vega/altair/issues/3610#issuecomment-2385432948)
2. Rename (make private) `alt.theme.themes` -> `alt.theme._themes` as mentioned in [#3610-comment-7](https://github.com/vega/altair/issues/3610#issuecomment-2385487732)
3. Embed multiple method examples into the `message`, as proposed by @mattijn in [#3618-comment](https://github.com/vega/altair/pull/3618#issuecomment-2408133664)
	- The main issue here is that even at the *current* length, [known usage](https://sourcegraph.com/search?q=context:global+alt.themes.+not+repo:altair+lang:Python+count:11000&patternType=keyword&sm=0) outputs a lot of content
		<details><summary>Screenshot</summary>
		<p>

		![image](https://github.com/user-attachments/assets/80fbfe30-eed3-4b52-bdb9-7936fedb0f76)

		</p></details> 
4. Add more granular deprecation warnings on the underlying `ThemeRegistry` methods
	- Something like
		- the top level `alt.themes` display links to the new **User Guide** section
		- `themes.register` has an example of the decorator transition
			- Or refer to the example in docstring
			- Or refer to some section of the **User Guide**
		- `themes.enable` is a trivial case, just specify `alternative`
	- I think there's more than one way this could be achieved
5. Display the warning *only* on the **first** access to `alt.themes`
	- There's an example of this in the [deprecated](https://github.com/search?q=repo%3Alaurent-laporte-pro%2Fdeprecated+%2Faction%5B%3D%3A%5C*%5D%2F&type=code) package
		- Seems to work by using `action="once"` with https://docs.python.org/3/library/warnings.html#the-warnings-filter
	- This option is more about mitigating effects of **3**
		- but *could* result in fewer users knowing the planned change

dangotbanned commented:
@mattijn just wanted to check if you had any thoughts on **3**-**5** in https://github.com/vega/altair/pull/3618#issuecomment-2408550858?

Currently planning to go ahead with **2** and then explore **4** some more today.

# Side note
I'm hoping to tackle #3497 once this one is merged, since
- we're now in [`3.8` EOL](https://devguide.python.org/versions/#unsupported-versions)
- `3.13` support for `pyarrow` seems to be around the corner, and is the current blocker for #3591 
- this PR is the last one open that could be a hassle with resolving merge conflicts from #3497 

```cmd
>>> ruff check . --config "target-version = 'py39'" --output-format "concise"
Found 86 errors.
[*] 70 fixable with the `--fix` option (2 hidden fixes can be enabled with the `--unsafe-fixes` option).
```

<details><summary>86 errors</summary>
<p>

```py
altair\expr\core.py:3:1: UP035 `typing.Dict` is deprecated, use `dict` instead
altair\expr\core.py:240:74: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\_dfi_types.py:10:1: UP035 [*] Import from `collections.abc` instead: `Iterable`
altair\utils\_show.py:5:1: UP035 [*] Import from `collections.abc` instead: `Iterable`
altair\utils\_transformed_data.py:3:1: UP035 [*] Import from `collections.abc` instead: `Iterable`
altair\utils\_transformed_data.py:3:1: UP035 `typing.Dict` is deprecated, use `dict` instead
altair\utils\_transformed_data.py:3:1: UP035 `typing.Tuple` is deprecated, use `tuple` instead
altair\utils\_transformed_data.py:37:20: UP006 [*] Use `tuple` instead of `Tuple` for type annotation
altair\utils\_transformed_data.py:38:27: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\_transformed_data.py:38:32: UP006 [*] Use `tuple` instead of `Tuple` for type annotation
altair\utils\_transformed_data.py:38:51: UP006 [*] Use `tuple` instead of `Tuple` for type annotation
altair\utils\_vegafusion_data.py:4:1: UP035 [*] Import from `collections.abc` instead: `MutableMapping`
altair\utils\compiler.py:1:1: UP035 `typing.Dict` is deprecated, use `dict` instead
altair\utils\compiler.py:8:34: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\compiler.py:8:51: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\compiler.py:11:69: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\core.py:15:1: UP035 [*] Import from `collections.abc` instead: `Iterator`
altair\utils\data.py:9:1: UP035 [*] Import from `collections.abc` instead: `MutableMapping`, `Sequence`
altair\utils\data.py:9:1: UP035 `typing.Dict` is deprecated, use `dict` instead
altair\utils\data.py:9:1: UP035 `typing.List` is deprecated, use `list` instead
altair\utils\data.py:57:5: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\data.py:63:31: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\data.py:64:21: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\data.py:64:37: UP006 [*] Use `list` instead of `List` for type annotation
altair\utils\data.py:64:42: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\data.py:66:33: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\data.py:66:49: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\data.py:66:65: UP006 [*] Use `list` instead of `List` for type annotation
altair\utils\data.py:66:70: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\data.py:67:41: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\display.py:7:1: UP035 `typing.Dict` is deprecated, use `dict` instead
altair\utils\display.py:7:1: UP035 `typing.Tuple` is deprecated, use `tuple` instead
altair\utils\display.py:21:33: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\display.py:22:37: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\display.py:24:25: UP006 [*] Use `tuple` instead of `Tuple` for type annotation
altair\utils\display.py:29:40: UP006 [*] Use `tuple` instead of `Tuple` for type annotation
altair\utils\display.py:30:5: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\display.py:30:26: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\display.py:30:44: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\display.py:30:54: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\schemapi.py:16:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`, `Mapping`, `Sequence`
altair\utils\schemapi.py:16:1: UP035 `typing.Dict` is deprecated, use `dict` instead
altair\utils\schemapi.py:16:1: UP035 `typing.List` is deprecated, use `list` instead
altair\utils\schemapi.py:70:34: UP006 [*] Use `list` instead of `List` for type annotation
altair\utils\schemapi.py:71:38: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\schemapi.py:1354:46: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\utils\schemapi.py:1354:62: UP006 [*] Use `list` instead of `List` for type annotation
altair\utils\selection.py:4:1: UP035 `typing.Dict` is deprecated, use `dict` instead
altair\utils\selection.py:4:1: UP035 `typing.List` is deprecated, use `list` instead
altair\utils\selection.py:8:26: UP006 [*] Use `list` instead of `List` for type annotation
altair\utils\selection.py:8:31: UP006 [*] Use `dict` instead of `Dict` for type annotation
altair\vegalite\v5\api.py:13:1: UP035 [*] Import from `collections.abc` instead: `Sequence`
altair\vegalite\v5\api.py:50:5: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`
altair\vegalite\v5\api.py:663:26: UP006 [*] Use `list` instead of `t.List` for type annotation
altair\vegalite\v5\schema\_config.py:7:1: UP035 [*] Import from `collections.abc` instead: `Sequence`
altair\vegalite\v5\schema\_typing.py:8:1: UP035 [*] Import from `collections.abc` instead: `Mapping`, `Sequence`
altair\vegalite\v5\schema\_typing.py:35:4: UP036 Version block is outdated for minimum Python version
altair\vegalite\v5\schema\_typing.py:38:5: UP035 [*] Import from `typing` instead: `Annotated`
altair\vegalite\v5\schema\channels.py:14:1: UP035 [*] Import from `collections.abc` instead: `Sequence`
altair\vegalite\v5\schema\core.py:8:1: UP035 [*] Import from `collections.abc` instead: `Iterator`, `Sequence`
altair\vegalite\v5\schema\mixins.py:6:1: UP035 [*] Import from `collections.abc` instead: `Sequence`
sphinxext\code_ref.py:17:5: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`, `Mapping`, `Sequence`
sphinxext\schematable.py:6:1: UP035 [*] Import from `collections.abc` instead: `Iterator`, `Sequence`
tests\__init__.py:15:5: UP035 [*] Import from `collections.abc` instead: `Collection`, `Iterator`, `Mapping`
tests\expr\test_expr.py:6:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`
tests\utils\test_schemapi.py:13:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Sequence`
tools\codemod.py:12:1: UP035 [*] Import from `collections.abc` instead: `Iterable`
tools\codemod.py:25:5: UP035 [*] Import from `collections.abc` instead: `Iterator`
tools\codemod.py:59:4: UP036 Version block is outdated for minimum Python version
tools\generate_api_docs.py:8:1: UP035 [*] Import from `collections.abc` instead: `Iterator`
tools\generate_schema_wrapper.py:15:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`
tools\markup.py:8:1: UP035 [*] Import from `collections.abc` instead: `Iterable`
tools\schemapi\codegen.py:11:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`
tools\schemapi\schemapi.py:14:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`, `Mapping`, `Sequence`
tools\schemapi\schemapi.py:14:1: UP035 `typing.Dict` is deprecated, use `dict` instead
tools\schemapi\schemapi.py:14:1: UP035 `typing.List` is deprecated, use `list` instead
tools\schemapi\schemapi.py:68:34: UP006 [*] Use `list` instead of `List` for type annotation
tools\schemapi\schemapi.py:69:38: UP006 [*] Use `dict` instead of `Dict` for type annotation
tools\schemapi\schemapi.py:1352:46: UP006 [*] Use `dict` instead of `Dict` for type annotation
tools\schemapi\schemapi.py:1352:62: UP006 [*] Use `list` instead of `List` for type annotation
tools\schemapi\utils.py:13:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`, `Mapping`, `Sequence`
tools\update_init_file.py:11:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`
tools\update_init_file.py:21:5: UP006 [*] Use `list` instead of `t.List` for type annotation
tools\update_init_file.py:22:5: UP006 [*] Use `dict` instead of `t.Dict` for type annotation
tools\update_init_file.py:23:5: UP006 [*] Use `tuple` instead of `t.Tuple` for type annotation
tools\vega_expr.py:19:1: UP035 [*] Import from `collections.abc` instead: `Iterable`, `Iterator`, `Mapping`, `Sequence`
```

</p>
</details> 

```cmd
>>> ruff check . --config "target-version = 'py39'" --statistics
44      UP035   [ ] deprecated-import
40      UP006   [*] non-pep585-annotation
 2      UP036   [*] outdated-version-block
```

So it seems like a good time to me to get that ready for `5.5.0` as planned it https://github.com/vega/altair/issues/3497#issuecomment-2258964446

mattijn commented:
I'm OK with option 5

dangotbanned commented:
> I'm OK with option 5 (from choices 3-5). Option 2 is good as well.

@mattijn thanks for the quick response

I'm having a hard time getting **5** working

> There's an example of this in the [deprecated](https://github.com/search?q=repo%3Alaurent-laporte-pro%2Fdeprecated+%2Faction%5B%3D%3A%5C*%5D%2F&type=code) package

If I've understood https://github.com/laurent-laporte-pro/deprecated/issues/64#issuecomment-1565599108 correctly, this approach doesn't work for `action="once"`

---

It isn't a huge job to track (ourselves) if we've served this warning yet. 
However, it would require adding some global state - which might not be ideal

dangotbanned commented:
@mattijn, as of [docs: Add examples to deprecation message](https://github.com/vega/altair/pull/3618/commits/5a1dc6893134966635999fb7f687316f1f10b5dd), the following message is displayed on first access **only** to `alt.themes`:

<details><summary>Warning output</summary>
<p>

```
/Temp/ipykernel_9448/3158091034.py:3: AltairDeprecationWarning: Deprecated in `altair=5.5.0`. Use altair.theme instead.
Most cases require only the following change:

    # Deprecated
    alt.themes.enable('quartz')

    # Updated
    alt.theme.enable('quartz')

If your code registers a theme, make the following change:

    # Deprecated
    def custom_theme():
        return {'height': 400, 'width': 700}
    alt.themes.register('theme_name', custom_theme)
    alt.themes.enable('theme_name')

    # Updated
    @alt.theme.register('theme_name', enable=True)
    def custom_theme() -> alt.theme.ThemeConfig:
        return {'height': 400, 'width': 700}

See the updated User Guide for further details:
    https://altair-viz.github.io/user_guide/api.html#theme
    https://altair-viz.github.io/user_guide/customization.html#chart-themes
  alt.themes

```


</p>
</details> 

Comparing to (https://github.com/vega/altair/pull/3618#issuecomment-2408133664), the main change was removing:

```py
import altair as alt
```

It isn't possible to trigger the warning *without* this import having occurred.
The final line of the warning will point to exactly **where** accessing the `alt` namespace happened.

---

I've got one last thing I'm experimenting with locally, that I hadn't thought of earlier.

Essentially, option **4** from (https://github.com/vega/altair/pull/3618#issuecomment-2408550858) with two tweaks:
- Only for `alt.themes.register`
- Using `deprecated(..., category=None)` to [avoid adding a runtime warning](https://docs.python.org/3/library/warnings.html#warnings.deprecated)

<details><summary>Default visual feedback</summary>
<p>

![image](https://github.com/user-attachments/assets/2b1b311f-266c-4c10-868e-571565a2ffde)

</p>
</details> 

This would be **in addition to** the runtime warning for `alt.themes`.
However, it would mean static-typing users could very clearly see **all** instances they've used it.

Especially when opting-in to `pyright`'s [reportDeprecated](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportDeprecated) rule

<details><summary>[reportDeprecated] feedback</summary>
<p>


![image](https://github.com/user-attachments/assets/210cb505-9f96-4782-86ef-758f7860c203)

</p>
</details> 

Right now I'm working on getting deprecation message to pass through for this case, but still keep the versioning requirement from #3455

If we were to use this, I'd want the message to be more specific than:
> The method "register" in class "ThemeRegistry" is deprecated

mattijn commented:
Great! One simple question. 
Can we add a new line (`\n`) to the first line of the warning? Currently the core message is a bit hidden at the end of a very long line.

From current:
```cmd
/Temp/ipykernel_9448/3158091034.py:3: AltairDeprecationWarning: Deprecated in `altair=5.5.0`. Use altair.theme instead.
```

To wish:
```
/Temp/ipykernel_9448/3158091034.py:3: 
AltairDeprecationWarning: Deprecated in `altair=5.5.0`. Use altair.theme instead.
```

dangotbanned commented:
> Great! One simple question. Can we add a new line (`\n`) to the first line of the warning? Currently the core message is a bit hidden at the end of a very long line.
> 
> From current:
> 
> ```batchfile
> /Temp/ipykernel_9448/3158091034.py:3: AltairDeprecationWarning: Deprecated in `altair=5.5.0`. Use altair.theme instead.
> ```
> 
> To wish:
> 
> ```
> /Temp/ipykernel_9448/3158091034.py:3: 
> AltairDeprecationWarning: Deprecated in `altair=5.5.0`. Use altair.theme instead.
> ```

@mattijn I would definitely prefer that, but the `AltairDeprecationWarning: ` part is added by [warnings.warn](https://docs.python.org/3/library/warnings.html#warnings.warn) itself.

AFAIK, the best we could do is:

```
/Temp/ipykernel_9448/3158091034.py:3: AltairDeprecationWarning: 
Deprecated in `altair=5.5.0`. Use altair.theme instead.
...
```

I'm open to changing the format for the part we can control:

https://github.com/vega/altair/blob/0245e0f4ebfb0ffa86043afa5cd1b4c9d6fe8a63/altair/utils/deprecation.py#L23-L32

mattijn commented:
That's also good👍

dangotbanned commented:
> That's also good👍

Fixed in (https://github.com/vega/altair/pull/3618/commits/22d7df2597704defca3360c1ca00eab18efdca55)

Really botched the commit message there, it should be:

`docs: Prefix "\n" to all deprecation warnings`

dangotbanned commented:
# Before
See also https://github.com/vega/altair/pull/3618#issuecomment-2423991968

![image](https://github.com/user-attachments/assets/210cb505-9f96-4782-86ef-758f7860c203)

# After (https://github.com/vega/altair/pull/3618/commits/25cbdc8119c867f7c74f52568245735083078c3f)

![image](https://github.com/user-attachments/assets/a4834f9e-ff7a-44d8-ba56-be2d1281f616)

mattijn commented:
Open question, you mention in the deprecation warning: `deprecated in v5.5.0`, would it make sense to change this into `deprecated **since** v5.5.0`?

dangotbanned commented:
> Open question, you mention in the deprecation warning: `deprecated in v5.5.0`, would it make sense to change this into `deprecated since v5.5.0`?

Yeah that makes sense @mattijn 

I hadn't noticed the convention before, but you can see it here https://docs.python.org/3/library/importlib.html#module-importlib.machinery

- Added in ...
- Changed in ...
- Deprecated since ...

I'm happy for you to make the change, or I'll update it tomorrow?

dangotbanned commented:
@mattijn I've addressed (https://github.com/vega/altair/pull/3618#issuecomment-2424199204) in (https://github.com/vega/altair/pull/3618/commits/a0497e0a5461274cceb06ffd26bf92f9d2ce60fe)

All of the [Unresolved conversations](https://github.com/vega/altair/pull/3618#issuecomment-2424199204) are just to help make the review easier.
You can dismiss once you've seen them to unblock merging

mattijn reviewed (approved):
Thanks, I approve. However, I’d like another maintainer to review this as well due to the significant impact of this PR. Once merged, it will require updates to several published packages on PyPI, such as those listed here: https://pypi.org/search/?q=altair+theme.

dangotbanned commented:
> Thanks, I approve. However, I’d like another maintainer to review this as well due to the significant impact of this PR. Once merged, it will require updates to several published packages on PyPI, such as those listed here: [pypi.org/search?q=altair+theme](https://pypi.org/search/?q=altair+theme).

Thanks for the review @mattijn!

Yeah that makes sense, always good to have another set of eyes

dangotbanned commented:
> Thanks, I approve. However, I’d like another maintainer to review this as well due to the significant impact of this PR. Once merged, it will require updates to several published packages on PyPI, such as those listed here: [pypi.org/search?q=altair+theme](https://pypi.org/search/?q=altair+theme).

@mattijn @jonmmease 

Apologies if this was already clear, but I just wanted to restate it in a short format for some reassurance.

---

- Merging this PR will require **no changes to dependent packages** using the public API of `altair` until at least `6.0.0`.
- Using `alt.themes` in `>=5.5.0; <6.0.0` will display a **[warning](https://github.com/vega/altair/pull/3618#issuecomment-2423991968)** on first use **only**
- **Static typing users** (*who have opted-in to this kind of reporting*) will see [warnings displayed on every use](https://github.com/vega/altair/pull/3618#issuecomment-2424101617) of `alt.themes.register`

dangotbanned commented:
Caught that one quite late 😅 (https://github.com/vega/altair/pull/3618#event-14991308294)

mattijn commented:
Any opinion on this @joelostblom, @dangotbanned? I'm not a themes/theme user myself, so it is hard for me to overview the impact. Do we know anyone who is/are using the theme and will be impacted by this?

mattijn commented:
Or, put positively, someone who could benefit from this PR and might be able to provide a quick check. Otherwise I will press the merge button in a week, November 11.
